core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -66,6 +66,21 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
       Future.successful(snapshotRef)
     }
 
+  def cloneSnapshotByReference(sourceWorkspaceId: UUID,
+                               targetWorkspaceId: UUID,
+                               snapshot: DataRepoSnapshotResource
+  ): DataRepoSnapshotResource = {
+    if (!workspaceStubExists(targetWorkspaceId, ctx)) {
+      workspaceManagerDAO.createWorkspace(targetWorkspaceId, ctx)
+    }
```

Comment:
As a standalone method, this is fine. But in the current PR, `WorkspaceService` may retrieve tens or hundreds of references, then call `cloneSnapshotByReference` for each one - which means we'll execute `workspaceStubExists` over and over. Each call to `workspaceStubExists` is a new REST request to WSM, so this is pretty inefficient.

I also commented elsewhere in the PR - I suggest `SnapshotService` should have a "clone all references" method. In that method, you could call `workspaceStubExists` and `createWorkspace` (if necessary) once at the top of the method, then retrieve and clone a whole bunch of references.